Add daemon option `--default-shm-size`

yongtang opened the pull request:
**- What I did**

This fix fixes issue raised in #29492 where it was not possible to specify a default `--default-shm-size` in daemon configuration for each `docker run``.

**- How I did it**

The flag `--default-shm-size` which is reloadable, has been added to the daemon configuation.
Related docs has been updated.

**- How to verify it**

**- Description for the changelog**

Add daemon option `--default-shm-size` for default shm size of created containers.

**- A picture of a cute animal (not mandatory but encouraged)**

![u 968902900 274083092 fm 15 gp 0](https://cloud.githubusercontent.com/assets/6932348/21472098/555c2b7e-ca7e-11e6-8416-56bc6cef4678.jpg)


This fix fixes #29492.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add daemon option --default-shm-size

This fix fixes issue raised in 29492 where it was not
possible to specify a default `--default-shm-size` in daemon
configuration for each `docker run``.

The flag `--default-shm-size` which is reloadable, has been
added to the daemon configuation.
Related docs has been updated.

This fix fixes 29492.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
/cc @flx42 @justincormack @thaJeztah

flx42 commented:
Thanks @yongtang, I barely had time to look at how to implement it, I'm glad you did a PR.

flx42 commented on daemon/config_unix.go:
```diff
@@ -89,6 +91,7 @@ func (config *Config) InstallFlags(flags *pflag.FlagSet) {
 	flags.Int64Var(&config.CPURealtimePeriod, "cpu-rt-period", 0, "Limit the CPU real-time period in microseconds")
 	flags.Int64Var(&config.CPURealtimeRuntime, "cpu-rt-runtime", 0, "Limit the CPU real-time runtime in microseconds")
 	flags.StringVar(&config.SeccompProfile, "seccomp-profile", "", "Path to seccomp profile")
+	flags.Int64Var(&config.ShmSize, "shm-size", defaultShmSize, "Set the default shm size for containers")
```
So, should it be `default-shm-size`? I think it's better, otherwise it could also mean the shared memory size for the docker daemon itself. So maybe `default-shm-size` is a better option, despite the concern raised by @thaJeztah regarding the proliferation of `default-*` options.

yongtang replied:
Thanks @flx42. The PR has been updated to use `default-shm-size` now.

flx42 commented on docs/reference/commandline/dockerd.md:
```diff
@@ -75,7 +75,8 @@ Options:
       --seccomp-profile value                 Path to seccomp profile
       --selinux-enabled                       Enable selinux support
-      --shutdown-timeout=15                   Set the shutdown timeout value in seconds
+      --shm-size int                          Set the default shm size for containers (default 67108864)
```
Does that mean you can't use units with this new option? When using `docker run`, you can, so ideally it should be the same here. For instance: `--shm-size=128m`

yongtang replied:
@flx42 The human readable memory size (like `128m`) would be tricky for daemon config.

Currently the processing of `docker run` and `dockerd` are different. The `dockerd` could either takes a flag like `--default-shm-size=67108864` or allows the user to specify it in a json file like:
```
{
"default-shm-size" : 67108864
}
```

The flag form in `dockerd` could be processed like `docker run` to take a human readable memory size. 

However, `dockerd` will directly unmarshal the json config file into a struct `Config`. As a result, if we use string (for human readable memory size) for `default-shm-size`, it is not possible to convert it into `int` and store it in `daemon.configStore` (unless we change the current way of storing config in docker daemon).

I think such a change probably could be done in a separate PR as it involves much bigger impact. Would also like to be confirmed by docker maintainers before making the change.

flx42 commented on man/dockerd.8.md:
```diff
@@ -321,6 +322,10 @@ unix://[/path/to/socket] to use.
 
+**--shm-size**=*67108864*
+  Set the daemon-wide default shm size for containers. Default is `67108864` (`64M`).
+  Set the shutdown timeout value in seconds. Default is `15`.
```
Copy/paste fail :)

yongtang replied:
Thanks @flx42. Must be due to the holiday mode. 😅 The PR has been updated now.

AkihiroSuda commented:
CI failed
```
00:07:17.861 --- FAIL: TestAdjustCPUShares (0.00s)
00:07:17.861 panic: runtime error: invalid memory address or nil pointer dereference [recovered]
00:07:17.861 	panic: runtime error: invalid memory address or nil pointer dereference
00:07:17.861 [signal SIGSEGV: segmentation violation code=0x1 addr=0x3b8 pc=0x4bf1b0]
00:07:17.862 
00:07:17.862 goroutine 51 [running]:
00:07:17.862 panic(0x13e5e20, 0xc4200100a0)
00:07:17.862 	/usr/local/go/src/runtime/panic.go:500 +0x1a1
00:07:17.862 testing.tRunner.func1(0xc42033ef00)
00:07:17.862 	/usr/local/go/src/testing/testing.go:579 +0x25d
00:07:17.862 panic(0x13e5e20, 0xc4200100a0)
00:07:17.862 	/usr/local/go/src/runtime/panic.go:458 +0x243
00:07:17.862 github.com/docker/docker/daemon.(*Daemon).adaptContainerSettings(0xc420213998, 0xc420213b80, 0xc42032f201, 0x26, 0xc42032f200)
00:07:17.863 	github.com/docker/docker/daemon/_test/_obj_test/daemon_unix.go:319 +0x130
00:07:17.863 github.com/docker/docker/daemon.TestAdjustCPUShares(0xc42033ef00)
00:07:17.863 	/go/src/github.com/docker/docker/daemon/daemon_unix_test.go:34 +0x177
00:07:17.863 testing.tRunner(0xc42033ef00, 0x1696ee0)
00:07:17.864 	/usr/local/go/src/testing/testing.go:610 +0x81
00:07:17.864 created by testing.(*T).Run
00:07:17.864 	/usr/local/go/src/testing/testing.go:646 +0x2ec
00:07:17.864 FAIL	github.com/docker/docker/daemon	2.044s
```

yongtang commented:
Thanks @AkihiroSuda. The PR has been updated and the CI failure has been fixed.

@flx42 I realized that we probably could have a customized JSON unmarshaler for human readable memory size parsing (e.g., `1g`, `128M`, etc). With a customized JSON unmarshaler there is no need to make dramatic changes in the current implementation of `daemon.configStore`.

The PR has been updated with the customized JSON unmarshaler and should take care of the human readable memory size issue. Please take a look and let me know if there are any issues.

thaJeztah commented on daemon/config_unix.go:
```diff
@@ -89,6 +95,7 @@ func (config *Config) InstallFlags(flags *pflag.FlagSet) {
 	flags.Int64Var(&config.CPURealtimePeriod, "cpu-rt-period", 0, "Limit the CPU real-time period in microseconds")
 	flags.Int64Var(&config.CPURealtimeRuntime, "cpu-rt-runtime", 0, "Limit the CPU real-time runtime in microseconds")
 	flags.StringVar(&config.SeccompProfile, "seccomp-profile", "", "Path to seccomp profile")
+	flags.Var(&config.ShmSize, "default-shm-size", "Set the default shm size for containers")
```
Perhaps `Default shm size for container` to be more consistent with other options

yongtang replied:
Thanks @thaJeztah. Done.

thaJeztah commented:
We should try to get `dockerd` and `docker run` consistent;

- `docker run` uses "string", not "bytes"
- `docker run` shows 64 MB, not MiB

Not sure which one is "best" though :smile:

On `docker run`;

    --shm-size string               Size of /dev/shm, default value is 64MB
 
 
On `dockerd`; 

    --default-shm-size bytes        Set the default shm size for containers (default 64 MiB)

yongtang committed:
Update opts.MemBytes to disable default, and move `docker run/create/build` to use opts.MemBytes

This fix made several updates:
1. Update opts.MemBytes so that default value will not show up.
   The reason is that in case a default value is decided by daemon,
   instead of client, we actually want to not show default value.
2. Move `docker run/create/build` to use opts.MemBytes for `--shm-size`
   This is to bring consistency between daemon and docker run
3. docs updates.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @thaJeztah for the review. I made some updates to bring consistency between `docker run` and `dockerd`:
1. The `docker run --shm-size` (and `docker create/build --shm-size` as well) has been changed to use `opts.MemBytes`, thus the output type would be `bytes` (instead of `string`).
   The reason to change `docker run` (instead of `dockerd`) is that applying the logic of parsing `string` is going to be quite challenging in `dockerd` when combined with `config.json`. I also think `bytes` might be a better term as the input is much clear.
   *Open to suggestions if the other way is preferred.*
2. The default value of `--shm-size` in `docker run/create/build` is hidden now. The default value of `dockerd` is shown as 64M in the daemon.
   I realized that since we set a default value in the daemon, the client side default value actually depends on the setting of the daemon (a `0` is passed to API anyway).
   Therefore, the old default value in client side (`64M`) is actually irrelevant because client will only pass `0` and let daemon decide in old and new behavior. 
   Because of the above, we may want to hide the default value as it is not possible to know what the default value is at the client side.

Some related docs has been updated. As the above is quite a change that impact several places, I created a separate commit in this PR. If we prefer not to move forward with the above, I could strip the second commit from the PR.

Please take a look and let me know if there are any issues.

yongtang commented:
Rebased to fix merging conflict.

LK4D4 commented:
@thaJeztah design lgty? flag sounds ok for me

yongtang commented:
Rebased to fix conflict.

LK4D4 commented:
ping @docker/core-maintainers

vdemeester reviewed (commented):
Design SGTM 🐯

yongtang commented:
@vdemeester The PR has been rebased and now all tests passed on Jenkins.

LK4D4 commented:
LGTM
ping @docker/core-maintainers

vdemeester reviewed (approved):
LGTM 🐸
/cc @thaJeztah for docs-review

vdemeester reviewed (approved):
docs LGTM 🐸
/cc @thaJeztah for docs revisit

thaJeztah commented:
/cc @albers @sdurrheimer for completion scripts

stowns commented:
this is listed as a valid arg for the daemon.json https://docs.docker.com/engine/reference/commandline/dockerd/.   
And it's listed as being included in the 17.04.0ce release notes https://docs.docker.com/release-notes/docker-ce/#17040-ce-2017-04-05. However, attempting to add it via Docker for Mac 17.09.0-ce results in 

![screen shot 2017-12-01 at 2 36 40 pm](https://user-images.githubusercontent.com/379774/33502372-65298d74-d6a5-11e7-9da6-269818089e77.png)

![screen shot 2017-12-01 at 2 39 46 pm](https://user-images.githubusercontent.com/379774/33502395-83f65994-d6a5-11e7-9dab-01b6a5df0542.png)

thaJeztah commented:
@stowns can you open an issue for that in the Docker for Mac issue tracker? The Docker for Mac UI has its own validation of the daemon.json file, and it looks like it wasn't updated to recognise the new option. The issue tracker for that is here: https://github.com/docker/for-mac/issues

To work around that problem, you can edit the `daemon.json` directly: on Docker for Mac, you can find the file in `~/.docker/daemon.json`

stowns commented:
@thaJeztah done https://github.com/docker/for-mac/issues/2338
